### PR TITLE
[CodeCompletion] Duplicate existential requirements when restated

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
Resolves [SR-7035](https://bugs.swift.org/browse/SR-7035).

Also resolves duplicates in cases like
```swift
func foo<T: A & B>(...) -> T.#^PARAM^#

class Foo : A, B {
  #^MEMBER^#
}
```

Please look into how I tell apart restated requirements. If there is a better way (ideally wipe out any `self` occurences from a type, which isn't trivial as far as I see), I'll be glad to hear.

I would like to know if it's more efficient and better to do 
```swift
if (decl.getKind() >= DeclKind::AbstractFuncDecl_First &&
    decl.getKind() <= DeclKind::AbstractFuncDecl_Last)
```
than `isa<AbstractFunctionDecl>(decl)`.

**P.S.** There is something strange going on with the `SemanticContextKind` on the tests. Have a look, it can be `Super` or `CurrNominal` depending on whether the function argument is or isn't of an archetype type. This was already present, the tests simply checked for one of them among the duplicates.

### CODE REVIEW

File: lib/AST/LookupVisibleDecls.cpp [hunk 1]
```diff
@@ -473,19 +473,70 @@ lookupVisibleMemberDeclsImpl(Type BaseTy, VisibleDeclConsumer &Consumer,
...
+      }
+      return;
+    }
+    auto type =
+      VD->getOverloadSignatureType()->getAs<AnyFunctionType>()->getResult();
+
+    if (foundFuncs[{VD->getFullName(), type}].empty()) {
+      parentConsumer.foundDecl(VD, Reason);
+    }
+    foundFuncs[{VD->getFullName(), type}].insert(VD);
```
Comment (slavapestov) [inline lib/AST/LookupVisibleDecls.cpp]: The return value of insert() tells you if the entry was inserted or not so you can use that instead.

Why is the value of this map a set of types here, do you actually do anything with them?

File: lib/AST/LookupVisibleDecls.cpp [hunk 2]
```diff
@@ -473,19 +473,70 @@ lookupVisibleMemberDeclsImpl(Type BaseTy, VisibleDeclConsumer &Consumer,
...
+  RestateFilteringConsumer(VisibleDeclConsumer &parentConsumer,
+                           Type baseTy, const DeclContext *DC,
+                           LazyResolver *resolver)
+  : parentConsumer(parentConsumer), baseTy(baseTy), DC(DC), resolver(resolver) {
+    validate();
+  }
+
+  void foundDecl(ValueDecl *VD, DeclVisibilityKind Reason) override {
+    assert(VD);
+    if (!VD->getDeclContext()->getAsProtocolOrProtocolExtensionContext()) {
```
Comment (slavapestov) [inline lib/AST/LookupVisibleDecls.cpp]: I don’t understand why this is specific to protocol extensions

File: lib/AST/LookupVisibleDecls.cpp [hunk 3]
```diff
@@ -473,19 +473,70 @@ lookupVisibleMemberDeclsImpl(Type BaseTy, VisibleDeclConsumer &Consumer,
...
+      return;
+    }
+    if (!VD->getInterfaceType()->is<AnyFunctionType>()) {
+      if (foundVars.insert(VD->getFullName()).second) {
+        parentConsumer.foundDecl(VD, Reason);
+      }
+      return;
+    }
+    auto type =
+      VD->getOverloadSignatureType()->getAs<AnyFunctionType>()->getResult();
```
Comment (slavapestov) [inline lib/AST/LookupVisibleDecls.cpp]: getAs returns null if it’s the wrong type; use castTo() if you don’t expect it to fail

File: lib/AST/LookupVisibleDecls.cpp [hunk 4]
```diff
@@ -551,13 +602,13 @@ static void lookupVisibleMemberDeclsImpl(
...
 
   // Enumerate members of archetype's requirements.
-  if (ArchetypeType *Archetype = BaseTy->getAs<ArchetypeType>()) {
-    for (auto Proto : Archetype->getConformsTo())
-      lookupVisibleProtocolMemberDecls(
-          BaseTy, Proto->getDeclaredType(), Consumer, CurrDC, LS,
-          getReasonForSuper(Reason), TypeResolver, GSB, Visited);
-
-    if (auto superclass = Archetype->getSuperclass())
+  if (ArchetypeType *AT = BaseTy->getAs<ArchetypeType>()) {
```
Comment (slavapestov) [inline lib/AST/LookupVisibleDecls.cpp]: Is this a non-functional change? Can you split it off into a separate patch if it’s unrelated?

**Review by slavapestov — COMMENTED** [review]
What happens if one requirement is overloaded and the only difference between the overloads is the generic signatures? I think calling getResult() on the overload signature will produce the same type then. Eg,

```
func f<T : P>(_: T)
func f<T : Q>(_: T)

**Comment by slavapestov** [comment]
Using isa is always better than checking the kind directly. I’m not sure what you mean with regard to wiping out Self requirements - can you give an example?

**Comment by AnthonyLatsis** [comment]
> What happens if one requirement is overloaded and the only difference between the overloads is the generic signatures? I think calling getResult() on the overload signature will produce the same type then.

I checked for that and surprisingly, it works. This must be due to the information `T` carries still being in the signature.
However, I am very curious about type comparing. It is done by pointer equality, but doesn't that mean that two equal but separate instances won't be equal? Sometimes it's useful to strip off information off types and compare them, but it works even then through pointers. How come?

**Comment by AnthonyLatsis** [comment]
> Using isa is always better than checking the kind directly. I’m not sure what you mean with regard to wiping out Self requirements - can you give an example?

Thank you for the advice. So it is safer, but less efficient. 
The problem is to tell whether this `ValueDecl` is a restatement,
```swift
protocol P {
  func foo()
}
protocol P1: P {
  func foo() // like here
}
```
The type of `foo` in `P` is `<Self where Self: P> (P)->()->()`. In `P1`, it's `<Self where Self: P1> (P1)->()->()`. This is simple, I can get the result type here. But how about a generic function that has both `Self` and `T`? Well, I do the same and it works, likely due to the archetype still being present in the signature. As you can see, the only thing that hiders us is the presence of `Self`.
> I’m not sure what you mean with regard to wiping out Self requirements - can you give an example?

That said, and considering subscripts have to be handled specially since they are not self-curried with the generic signature still floating around, ideally, to tell apart restatements we have to remove `Self` from the type, be it an associated type or a function. 

`<Self where Self: P, T>(P) -> (arg: T) -> ()   becomes `<T>(arg: T)->(). As you see, I do the same except for the inclusion of the left-over generic signature.

File: lib/AST/LookupVisibleDecls.cpp (hunk 2 above)
Comment (AnthonyLatsis) [reply lib/AST/LookupVisibleDecls.cpp]: It isn't! I'll change this to `dyn_cast<ProtocolDecl>`

File: lib/AST/LookupVisibleDecls.cpp (hunk 1 above)
Comment (AnthonyLatsis) [reply lib/AST/LookupVisibleDecls.cpp]: > The return value of insert() tells you if the entry was inserted or not so you can use that instead.

It's safer to check for emptyness, because I have a set a pointers. Actually, I don't want more that one element that is equal by value, so I better change the container to `TinyVector`.

**Comment by slavapestov** [comment]
Types are compared for pointer equality because they are memoized. We never allocate the same type twice. However, type aliases and other sugar do create unique types, so you should always use isEqual() (or compare CanTypes) to strip off the sugar before.

I think the reason the example with overloads by generic signature worked is because it was comparing sugared types. Each generic parameter type gets a sugared type with its name as written in source, but when canonicalized, only the index in the signature matters, so I think it will stop working once you use canonical types instead.

Eg, we want these two to be the same decl:

```
typealias X = Int
func f() -> X
func f() -> Int
```

**Comment by AnthonyLatsis** [comment]
Thank you for the explanation. I do allocate compute new signatures and create new types here for comparison reasons, but they still somehow manage to correctly compare if they are canonical. 
 
> so I think it will stop working once you use canonical types instead.

I faced an opposite problem when fixing the overload problem: two signatures, for example
```swift
<Self> (Self) -> () -> ()
<Self> (Self) -> () -> ()
```
compare as different if I do not canonicalize them and equal otherwise. Is this due to each `Self` being a different instance? If so, what changes after canonicalization that makes these two equal? The canonicalized version is `<τ_0_0> (τ_0_0) -> () -> ()`

**Comment by AnthonyLatsis** [comment]
I now only remove the requirements from `Self` when dealing with subscripts or explicitly generic functions. This fixes cases overload cases where the only difference is the generic signature.

**Comment by AnthonyLatsis** [comment]
Oops, conflicts from [#16534](https://github.com/apple/swift/pull/16534)

File: lib/AST/LookupVisibleDecls.cpp [hunk 5]
```diff
@@ -473,19 +473,86 @@ lookupVisibleMemberDeclsImpl(Type BaseTy, VisibleDeclConsumer &Consumer,
...
+    // from Self to make sure it doesn't prevent us from recognizing restatements.
+    CanType type;
+    // This can't be null since we are dealing with method or
+    // subscript requirements, where Self is an implicit generic param.
+    auto sig = AFT->getOptGenericSignature();
+    auto params = sig->getGenericParams();
+    if (params.size() == 1 && !isa<SubscriptDecl>(VD)) {
+      type = AFT->getResult()->getCanonicalType();
+    } else {
+      auto newReqs = sig->getRequirements().drop_front();
```
Comment (slavapestov) [inline lib/AST/LookupVisibleDecls.cpp]: Please don’t make assumptions about the order of requirements in the generic signature like this. There must be a better way to accomplish what you’re trying to do. Perhaps you want to drop any conformance requirement with Self on the left hand side instead of assuming it’s the first one. Also this kind of thing should be factored out into a new method; the code here is getting a bit too long.

**Comment by slavapestov** [comment]
Yes, each protocol has its own sugared Self type, but they all canonicalize to the same thing.

File: lib/AST/LookupVisibleDecls.cpp (hunk 5 above)
Comment (slavapestov) [reply lib/AST/LookupVisibleDecls.cpp]: It will be longer than 30 lines once you do the right thing here, and it's more about single responsibility per method than raw length.

Comment (AnthonyLatsis) [reply lib/AST/LookupVisibleDecls.cpp]: I see, you mean the type-computing logic.

File: lib/AST/LookupVisibleDecls.cpp [hunk 6]
```diff
@@ -473,19 +474,96 @@ lookupVisibleMemberDeclsImpl(Type BaseTy, VisibleDeclConsumer &Consumer,
...
+
+    // This can't be null since we are dealing with method or
+    // subscript requirements, where Self is an implicit generic param.
+    auto sig = AFT->getOptGenericSignature();
+    auto params = sig->getGenericParams();
+
+    if (params.size() == 1 && !isa<SubscriptDecl>(VD)) {
+      return AFT->getResult()->getCanonicalType();
+    }
+    GenericTypeParamType *SelfParam;
```
Comment (slavapestov) [inline lib/AST/LookupVisibleDecls.cpp]: You don't have to "search" for it. Instead use VD->getDeclContext()->getSelfInterfaceType()

File: lib/AST/LookupVisibleDecls.cpp [hunk 7]
```diff
@@ -473,19 +474,96 @@ lookupVisibleMemberDeclsImpl(Type BaseTy, VisibleDeclConsumer &Consumer,
...
+    SmallVector<Requirement, 4> newReqs;
+    llvm::for_each(sig->getRequirements(), [&](Requirement req) {
+      if (!SelfParam->isEqual(req.getFirstType()))
+        newReqs.push_back(req);
+    });
+    auto newSig = GenericSignature::get(params, newReqs, false);
+
+    return GenericFunctionType::get(newSig, AFT->getInput(),
+                                    AFT->getResult(), AFT->getExtInfo())
+    ->getCanonicalType(newSig);
```
Comment (slavapestov) [inline lib/AST/LookupVisibleDecls.cpp]: Indentation needs to be fixed, and you can use the no-argument form of getCanonicalType() instead.

File: lib/AST/LookupVisibleDecls.cpp [hunk 8]
```diff
@@ -473,19 +474,96 @@ lookupVisibleMemberDeclsImpl(Type BaseTy, VisibleDeclConsumer &Consumer,
...
+    }
+    auto type = stripSelfRequirementsIfNeeded(VD, AFT);
+    if (foundFuncs.insert({VD->getFullName(), type}).second) {
+      parentConsumer.foundDecl(VD, Reason);
+    }
+  }
+private:
+  CanType stripSelfRequirementsIfNeeded(ValueDecl *VD, AnyFunctionType *AFT) {
+    // Preserve the generic signature if this is a subscript, which are uncurried,
+    // or if we have genenic params other than Self. Otherwise, strip if off
```
Comment (slavapestov) [inline lib/AST/LookupVisibleDecls.cpp]: typo: "genenic"

Comment (slavapestov) [reply lib/AST/LookupVisibleDecls.cpp]: typo: "strip if off"

File: lib/AST/LookupVisibleDecls.cpp [hunk 9]
```diff
@@ -473,19 +474,96 @@ lookupVisibleMemberDeclsImpl(Type BaseTy, VisibleDeclConsumer &Consumer,
...
+    // and use the resultType of the curried function type.
+    // When preserving the generic signature, we remove the requirements
+    // from Self to make they don't prevent us from recognizing restatements.
+
+    // This can't be null since we are dealing with method or
```
Comment (slavapestov) [inline lib/AST/LookupVisibleDecls.cpp]: In that case, the function should take a GenericFunctionType instead of an AnyFunctionType, then this comment becomes unnecessary

File: lib/AST/LookupVisibleDecls.cpp (hunk 6 above)
Comment (AnthonyLatsis) [reply lib/AST/LookupVisibleDecls.cpp]: Thank you, that is exactly what I was looking for...in the wrong place.

**Comment by AnthonyLatsis** [comment]
@slavapestov have a couple of minutes?

**Comment by AnthonyLatsis** [comment]
I have the similar [SR-7758](https://bugs.swift.org/browse/SR-7758) waiting which is dependent on the new consumer here. Can I commit to this PR then to solve them together?

**Comment by AnthonyLatsis** [comment]
cc @nkcsgexi @benlangmuir

File: test/IDE/complete_override_access_control_protocol.swift [hunk 10]
```diff
@@ -168,18 +168,17 @@ public class TestPublicDE : ProtocolDPrivate, ProtocolEPublic {
...
 // TEST_PRIVATE_DE-DAG: Decl[InstanceMethod]/Super: func collidingGeneric<T>(x: T) {|}{{; name=.+$}}
 
-// TEST_INTERNAL_DE: Begin completions, 4 items
+// TEST_INTERNAL_DE: Begin completions, 2 items
 // TEST_INTERNAL_DE-DAG: Decl[InstanceMethod]/Super: func colliding() {|}{{; name=.+$}}
 // TEST_INTERNAL_DE-DAG: Decl[InstanceMethod]/Super: func collidingGeneric<T>(x: T) {|}{{; name=.+$}}
 
-// TEST_PUBLIC_DE: Begin completions, 4 items
-// TEST_PUBLIC_DE-DAG: Decl[InstanceMethod]/Super: public func colliding() {|}{{; name=.+$}}
-// TEST_PUBLIC_DE-DAG: Decl[InstanceMethod]/Super: public func collidingGeneric<T>(x: T) {|}{{; name=.+$}}
```
Comment (benlangmuir) [inline test/IDE/complete_override_access_control_protocol.swift]: Why did we lose the public keyword here?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by AnthonyLatsis** [comment]
I guess these are failures from the recently merged removal of `self` from type identifiers.

**Comment by AnthonyLatsis** [comment]
The failures where because some tests began having unstable result ordering and the `NEXT` checks fail from time to time. This is because while processing, I don't respect the order in which they would have been fed to the parent consumer. Is this something to be concerned about? Should I write additional logic to keep the original order? If just stability is important, the order can be made stable by using `MapVector`, but it will still be different from how it was. You can have a look at what changed to `DAG` in the diff..

**Comment by slavapestov** [comment]
@AnthonyLatsis Yes, order stability is important. Please use MapVector and similar structures to ensure it doesn't depend on hash values.

**Comment by AnthonyLatsis** [comment]
Order restored

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by AnthonyLatsis** [comment]
I got it, let's try again.

**Comment by AnthonyLatsis** [comment]
@slavapestov Can you smoke test please?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test
